File: src/main/java/frc/robot/commands/DriveToFuel.java
package frc.robot.commands;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.CommandScheduler;
import frc.robot.constants.FuelVisionConstants;
import frc.robot.subsystems.fuelVision.FuelVision;
import frc.robot.subsystems.intake.Intake;
import frc.robot.subsystems.swerve.Drive;
import frc.utils.ExtraMath;
import frc.utils.RectZone;
import org.littletonrobotics.junction.Logger;
import org.w3c.dom.css.Rect;

import java.util.Comparator;
import java.util.Map;
import java.util.Set;
import java.util.function.Supplier;

import static edu.wpi.first.units.Units.*;

public class DriveToFuel extends Command {

    Drive drive;
    FuelVision fuelVision;

    Command driveCommand;

    Supplier<RectZone> zone;

    double angleToTarg = 0;
    double fieldVX = 0;
    double fieldVY = 0;

    public DriveToFuel(Drive drive, FuelVision fuelVision, Supplier<RectZone> allowedZone) {
        this.zone = allowedZone;
        this.drive = drive;
        this.fuelVision = fuelVision;

        driveCommand = drive.rotationLock(
                () -> angleToTarg,
                () -> fieldVX,
                () -> fieldVY
        );
        CommandScheduler.getInstance().registerComposedCommands(driveCommand);

        addRequirements(drive, fuelVision);
    }

    @Override
    public void initialize() {
        driveCommand.initialize();
    }

    @Override
    public void execute() {
        Map<FuelVision.gridCoord, Set<FuelVision.fuelData>> map = fuelVision.getFuelMap();

        var best = map.entrySet().stream()
                .filter(entry -> zone.get().contains(new Translation2d(entry.getKey().x()* FuelVisionConstants.GRID_SIZE.in(Meters), entry.getKey().y()*FuelVisionConstants.GRID_SIZE.in(Meters))))
                .max(Comparator.comparingDouble(entry -> {
            int fuel = entry.getValue().size();
            double dist = new Translation2d(entry.getKey().x()* FuelVisionConstants.GRID_SIZE.in(Meters), entry.getKey().y()*FuelVisionConstants.GRID_SIZE.in(Meters)).getDistance(drive.getPose().getTranslation());
            return fuel/Math.max(dist, 0.01);
        }));
        if(best.isPresent() && !best.get().getValue().isEmpty()) {
            angleToTarg = ExtraMath.getAngleToPos(new Translation2d(best.get().getKey().x()* FuelVisionConstants.GRID_SIZE.in(Meters), best.get().getKey().y()*FuelVisionConstants.GRID_SIZE.in(Meters)), drive.getPose().getTranslation()).in(Radians);

            fieldVX = Math.cos(drive.getRotation().getRadians()-angleToTarg) * Math.cos(angleToTarg);
            fieldVY = Math.cos(drive.getRotation().getRadians()-angleToTarg) * Math.sin(angleToTarg);

            Logger.recordOutput("Auto/AI/target", new Translation2d(best.get().getKey().x()* FuelVisionConstants.GRID_SIZE.in(Meters), best.get().getKey().y()*FuelVisionConstants.GRID_SIZE.in(Meters)));
        } else {
            Logger.recordOutput("Auto/AI/target", ((Translation2d) null));
            angleToTarg = drive.getRotation().plus(Rotation2d.fromDegrees(90*0.02)).getRadians();
            fieldVX = 0;
            fieldVY = 0;
        }
        driveCommand.execute();
    }

    @Override
    public void end(boolean interrupted) {
        driveCommand.end(interrupted);
    }

    @Override
    public boolean isFinished() {
        return driveCommand.isFinished();
    }
}

File: src/main/kotlin/frc/robot/commands/DriveToFuel.kt
package frc.robot.commands

import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.math.geometry.Translation2d
import edu.wpi.first.units.Units.Meters
import edu.wpi.first.units.Units.Radians
import edu.wpi.first.wpilibj2.command.Command
import edu.wpi.first.wpilibj2.command.CommandScheduler
import frc.robot.constants.FuelVisionConstants
import frc.robot.subsystems.fuelVision.FuelVision
import frc.robot.subsystems.swerve.Drive
import frc.utils.ExtraMath
import frc.utils.RectZone
import org.littletonrobotics.junction.Logger
import java.util.Comparator
import java.util.function.DoubleSupplier
import java.util.function.Supplier

class DriveToFuel(
    private val drive: Drive,
    private val fuelVision: FuelVision,
    private val zone: Supplier<RectZone>
) : Command() {

    private val driveCommand: Command

    private var angleToTarg: Double = 0.0
    private var fieldVX: Double = 0.0
    private var fieldVY: Double = 0.0

    init {
        driveCommand = drive.rotationLock(
            DoubleSupplier { angleToTarg },
            DoubleSupplier { fieldVX },
            DoubleSupplier { fieldVY }
        )
        CommandScheduler.getInstance().registerComposedCommands(driveCommand)

        addRequirements(drive, fuelVision)
    }

    override fun initialize() {
        driveCommand.initialize()
    }

    override fun execute() {
        val map = fuelVision.getFuelMap()

        val best = map.entries.stream()
            .filter { entry ->
                zone.get().contains(
                    Translation2d(
                        entry.key.x() * FuelVisionConstants.GRID_SIZE.`in`(Meters),
                        entry.key.y() * FuelVisionConstants.GRID_SIZE.`in`(Meters)
                    )
                )
            }
            .max(
                Comparator.comparingDouble { entry ->
                    val fuel = entry.value.size.toDouble()
                    val dist = Translation2d(
                        entry.key.x() * FuelVisionConstants.GRID_SIZE.`in`(Meters),
                        entry.key.y() * FuelVisionConstants.GRID_SIZE.`in`(Meters)
                    ).getDistance(drive.pose.translation)
                    fuel / Math.max(dist, 0.01)
                }
            )

        if (best.isPresent && best.get().value.isNotEmpty()) {
            val target = Translation2d(
                best.get().key.x() * FuelVisionConstants.GRID_SIZE.`in`(Meters),
                best.get().key.y() * FuelVisionConstants.GRID_SIZE.`in`(Meters)
            )
            angleToTarg = ExtraMath.getAngleToPos(target, drive.pose.translation).`in`(Radians)

            fieldVX = Math.cos(drive.rotation.radians - angleToTarg) * Math.cos(angleToTarg)
            fieldVY = Math.cos(drive.rotation.radians - angleToTarg) * Math.sin(angleToTarg)

            Logger.recordOutput("Auto/AI/target", target)
        } else {
            Logger.recordOutput("Auto/AI/target", null as Translation2d?)
            angleToTarg = drive.rotation.plus(Rotation2d.fromDegrees(180 * 0.02)).radians
            fieldVX = 0.0
            fieldVY = 0.0
        }
        driveCommand.execute()
    }

    override fun end(interrupted: Boolean) {
        driveCommand.end(interrupted)
    }

    override fun isFinished(): Boolean {
        return driveCommand.isFinished
    }
}
